File: backend/models/Painting.js
const { DataTypes } = require('sequelize');

class Shape {
    constructor(data = {}) {
        this.id = data.id || null;
        this.type = data.type || null;
        this.x = data.x || 0;
        this.y = data.y || 0;
        this.timestamp = data.timestamp || new Date().toISOString();
    }

    static fromJSON(json) {
        return new Shape(json);
    }

    toJSON() {
        return {
            id: this.id,
            type: this.type,
            x: this.x,
            y: this.y,
            timestamp: this.timestamp
        };
    }

    isValid() {
        return this.id && this.type &&
            typeof this.x === 'number' &&
            typeof this.y === 'number' &&
            this.timestamp;
    }
}

class Metadata {
    constructor(data = {}) {
        this.totalShapes = data.totalShapes || 0;
        this.shapeTypes = data.shapeTypes || {};
        this.canvasSize = data.canvasSize || { width: 800, height: 400 };
    }

    static fromJSON(json) {
        return new Metadata(json);
    }

    toJSON() {
        return {
            totalShapes: this.totalShapes,
            shapeTypes: this.shapeTypes,
            canvasSize: this.canvasSize
        };
    }

    updateFromShapes(shapes) {
        this.totalShapes = shapes.length;
        this.shapeTypes = {};

        shapes.forEach(shape => {
            if (this.shapeTypes[shape.type]) {
                this.shapeTypes[shape.type]++;
            } else {
                this.shapeTypes[shape.type] = 1;
            }
        });
    }
}

class Painting {
    constructor(data = {}) {
        this.title = data.title || 'Untitled';
        this.timestamp = data.timestamp || new Date().toISOString();
        this.version = data.version || '1.0';
        this.shapes = [];
        this.metadata = new Metadata(data.metadata);

        if (data.shapes && Array.isArray(data.shapes)) {
            this.shapes = data.shapes.map(shapeData => Shape.fromJSON(shapeData));
        }

        this.metadata.updateFromShapes(this.shapes);
    }

    static fromJSON(json) {
        if (!json) return null;

        if (typeof json === 'string') {
            try {
                json = JSON.parse(json);
            } catch (error) {
                console.error('Error parsing JSON:', error);
                return null;
            }
        }

        return new Painting(json);
    }

    toJSON() {
        return {
            title: this.title,
            timestamp: this.timestamp,
            version: this.version,
            shapes: this.shapes.map(shape => shape.toJSON()),
            metadata: this.metadata.toJSON()
        };
    }

    toString() {
        return JSON.stringify(this.toJSON());
    }

    addShape(shapeData) {
        const shape = Shape.fromJSON(shapeData);
        if (shape.isValid()) {
            this.shapes.push(shape);
            this.metadata.updateFromShapes(this.shapes);
            return true;
        }
        return false;
    }

    removeShape(shapeId) {
        const index = this.shapes.findIndex(shape => shape.id === shapeId);
        if (index !== -1) {
            this.shapes.splice(index, 1);
            this.metadata.updateFromShapes(this.shapes);
            return true;
        }
        return false;
    }

    updateShape(shapeId, newData) {
        const shape = this.shapes.find(shape => shape.id === shapeId);
        if (shape) {
            Object.assign(shape, newData);
            this.metadata.updateFromShapes(this.shapes);
            return true;
        }
        return false;
    }

    getShape(shapeId) {
        return this.shapes.find(shape => shape.id === shapeId);
    }

    getShapesByType(type) {
        return this.shapes.filter(shape => shape.type === type);
    }

    isValid() {
        return this.title &&
            this.timestamp &&
            this.version &&
            Array.isArray(this.shapes) &&
            this.shapes.every(shape => shape.isValid());
    }

    updateTimestamp() {
        this.timestamp = new Date().toISOString();
    }

    clearShapes() {
        this.shapes = [];
        this.metadata.updateFromShapes(this.shapes);
    }

    getStatistics() {
        return {
            title: this.title,
            totalShapes: this.metadata.totalShapes,
            shapeTypes: this.metadata.shapeTypes,
            lastModified: this.timestamp,
            canvasSize: this.metadata.canvasSize
        };
    }
}

module.exports = {
    Painting,
    Shape,
    Metadata
};

File: java-backend/src/main/java/com/example/canvasdrawingapi/model/painting/Painting.java
package com.example.canvasdrawingapi.model.painting;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Painting {
    private String title;
    private String timestamp;
    private String version;
    private List<Shape> shapes;
    private Metadata metadata;
    private boolean valid;

    public Painting() {
        this.title = "Untitled";
        this.timestamp = Instant.now().toString();
        this.version = "1.0";
        this.shapes = new ArrayList<>();
        this.metadata = new Metadata();
    }

    public Painting(String title) {
        this();
        this.title = title;
    }

    public static Painting fromJson(String json) {
        if (json == null || json.trim().isEmpty()) {
            return null;
        }

        try {
            ObjectMapper mapper = new ObjectMapper();
            Painting painting = mapper.readValue(json, Painting.class);
            if (painting.shapes == null) {
                painting.shapes = new ArrayList<>();
            }
            if (painting.metadata == null) {
                painting.metadata = new Metadata();
            }
            painting.metadata.updateFromShapes(painting.shapes);
            return painting;
        } catch (JsonProcessingException e) {
            e.printStackTrace();
            return null;
        }
    }

    public String toJson() {
        try {
            ObjectMapper mapper = new ObjectMapper();
            return mapper.writeValueAsString(this);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
            return null;
        }
    }

    public boolean addShape(Shape shape) {
        if (shape != null && shape.isValid()) {
            this.shapes.add(shape);
            this.metadata.updateFromShapes(this.shapes);
            return true;
        }
        return false;
    }

    public boolean removeShape(String shapeId) {
        boolean removed = this.shapes.removeIf(shape -> shape.getId().equals(shapeId));
        if (removed) {
            this.metadata.updateFromShapes(this.shapes);
        }
        return removed;
    }

    public void clearShapes() {
        this.shapes.clear();
        this.metadata.updateFromShapes(this.shapes);
    }

    public void updateTimestamp() {
        this.timestamp = Instant.now().toString();
    }

    public boolean isValid() {
        return title != null && !title.trim().isEmpty() &&
               timestamp != null &&
               version != null &&
               shapes != null &&
               shapes.stream().allMatch(Shape::isValid);
    }

    public Map<String, Object> getStatistics() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("title", title);
        stats.put("totalShapes", metadata.getTotalShapes());
        stats.put("shapeTypes", metadata.getShapeTypes());
        stats.put("lastModified", timestamp);
        stats.put("canvasSize", metadata.getCanvasSize());
        return stats;
    }

    // Getters and Setters
    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(String timestamp) {
        this.timestamp = timestamp;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public List<Shape> getShapes() {
        return shapes;
    }

    public void setShapes(List<Shape> shapes) {
        this.shapes = shapes != null ? shapes : new ArrayList<>();
        this.metadata.updateFromShapes(this.shapes);
    }

    public Metadata getMetadata() {
        return metadata;
    }

    public void setMetadata(Metadata metadata) {
        this.metadata = metadata != null ? metadata : new Metadata();
    }
}
